verticapy/learn/causal_model.py

```diff
@@ -0,0 +1,343 @@
+# import
+from verticapy import vDataFrame, vDataFrameSQL
+from verticapy.toolbox import check_types
+import verticapy as vp
+import itertools
+from tqdm.auto import tqdm
+from time import time
+from typing import Union
+
+import warnings
+
+from ..stats.tools import conditional_chi_square, is_dir_path
+import numpy as np
+from verticapy import plot
+
+
+# ---#
+class CausalModel:
+	"""
+-------------------------------------------------
+Main Class for Causal Models.
+
+Parameters
+----------
+input_relation: [str, vDataframe]
+	Name of original relation
+	"""	
+ 
+	# ---#
+	def __init__(self, input_relation : Union[str, vDataFrame]):
+		check_types([("input_relation", input_relation, [str, vDataFrame])])
+		vdf = input_relation
+		if isinstance(vdf, str):
+			vdf = vDataFrameSQL(relation=vdf)
+
+		self.input_relation = vdf
+		self.vdf = vdf.copy()
+			
+		self.skeleton = None
+		self.sepsets = None
+		self.pdag : dict = None
+
+
+	# ---#
+	def reformat(self, input_columns:list = [], 
+					   exclude_columns:list = [], 
+					   max_unique: int = 5, 
+					   method:str = ""): 
+		"""
+		-----------------------------------------------------------
+		Reformats the input vDataframe to optimize Causal Discovery.
+		
+		Parameters
+		----------
+		input_columns: list, optional
+			Columns relevant to the Causal Network
+		exclude_columns: list, optional
+			Columns not relevant to the Causal Network
+		max_unique: int, optional
+			The maximum number of unqiue elements in any column.
+		method: str, optional
+			If max_unique is specified, the method is used to 
+			discretize the numeric elements in the columns.
+
+		Returns
+		-------
+		None
+		"""
+		
+		temp = self.input_relation.copy()
+		if isinstance(input_columns, str):
+			input_columns = [input_columns]
+		if isinstance(exclude_columns, str):
+			exclude_columns = [exclude_columns]
+		check_types(
+			[
+				("input_columns", input_columns, [list]),
+				("exclude_columns", exclude_columns, [list]),
+				("max_unique", max_unique, [int]),
+				("method", method, [str])
+			]
+		)
+		if input_columns != []:
+			temp.are_namecols_in(input_columns)
+		if exclude_columns != []:
+			temp.are_namecols_in(exclude_columns)
+		if method != "":
+			if method not in ["auto", "same_freq", "same_width", "smart", "topk"]:
+				raise ParameterError(f"{method} is not a valid method for vDataFrame.discretize.")
+
+		if input_columns != []:
+			remove_cols = set(temp._VERTICAPY_VARIABLES_['columns']) - set(input_columns)
+			temp.drop(remove_cols)
+
+		if exclude_columns != []:
+			remove_cols = set(exclude_columns) - set(temp._VERTICAPY_VARIABLES_['columns'])	
+			temp.drop(remove_cols)
+
+		temp.dropna()
+		for col in temp._VERTICAPY_VARIABLES_['columns']:
+			if temp[col].isnum():
```

Comment:
Is `max_unique` only applied to numerical columns? What about the categorical columns? Can they have number of categories bigger than `max_unique`?

Reply:
This mainly applies to categorical columns, because numerical columns can be discretized. If there are too many categories then running this function is exponentially slower. Because of the way the function is set up, it needs to calculate the chi2 score for each state of `z` and if `z` is four columns each having 100 categories that is 100,000,000 iterations for just one conditional independence test. So it can be more, but we should warn against it for performance purposes.

Reply:
@zacandcheese  please add the warning in your code